compio-io/src/write/ext.rs

```diff
@@ -0,0 +1,141 @@
+use compio_buf::{buf_try, BufResult, IntoInner, IoBuf, IoVectoredBuf};
+
+use crate::{AsyncWrite, AsyncWriteAt, IoResult};
+
+/// Shared code for write a scalar value into the underlying writer.
+macro_rules! write_scalar {
+    ($t:ty, $be:ident, $le:ident) => {
+        ::paste::paste! {
+            #[doc = concat!("Write a big endian `", stringify!($t), "` into the underlying writer.")]
+            async fn [< write_ $t >](&mut self, num: $t) -> IoResult<()> {
+                use ::compio_buf::{arrayvec::ArrayVec, BufResult};
+
+                const LEN: usize = ::std::mem::size_of::<$t>();
+                let BufResult(len, _) = self
+                    .write_all(ArrayVec::<u8, LEN>::from(num.$be()))
+                    .await;
+                assert_eq!(len?, LEN, "`write_all` returned unexpected length");
+                Ok(())
+            }
+
+            #[doc = concat!("Write a little endian `", stringify!($t), "` into the underlying writer.")]
+            async fn [< write_ $t _le >](&mut self, num: $t) -> IoResult<()> {
+                use ::compio_buf::{arrayvec::ArrayVec, BufResult};
+
+                const LEN: usize = ::std::mem::size_of::<$t>();
+                let BufResult(len, _) = self
+                    .write_all(ArrayVec::<u8, LEN>::from(num.$le()))
+                    .await;
+                assert_eq!(len?, LEN, "`write_all` returned unexpected length");
+                Ok(())
+            }
+        }
+    };
+}
+
+/// Shared code for loop writing until all contents are written.
+macro_rules! loop_write_all {
+    ($buf:ident, $len:expr, $needle:ident,loop $expr_expr:expr) => {
+        let len = $len;
+        let mut $needle = 0;
+
+        while $needle < len {
+            let n;
+            (n, $buf) = buf_try!($expr_expr.await.into_inner());
+            if n == 0 {
+                return BufResult(
+                    Err(::std::io::Error::new(
+                        ::std::io::ErrorKind::WriteZero,
+                        "failed to write whole buffer",
+                    )),
+                    $buf,
+                );
+            }
+            $needle += n;
+        }
+
+        return BufResult(Ok($needle), $buf);
+    };
+}
+
+pub trait AsyncWriteExt: AsyncWrite {
+    /// Creates a "by reference" adaptor for this instance of [`AsyncWrite`].
+    ///
+    /// The returned adapter also implements [`AsyncWrite`] and will simply
+    /// borrow this current writer.
+    fn by_ref(&mut self) -> &mut Self
```

Comment:
Why do we need this?

Reply:
Because std has it

Reply:
What I'm implying is that, it's the same rationale: we implement the trait for both T and &mut T, `reader.by_ref().read` has better ergonomic than `(&mut reader).read()`, and the semantic is clearer (we explicitly saying we are using the reader/weiter by_ref)

Reply:
Fair enough, thanks.